### DISCUSSION TITLE
Why am I getting "Too many re-renders" when trying to setState?

### DISCUSSION THREAD

**Discussion by darrylmorley** [vercel/next.js#21060] [Help] [answered]
Hi,

Could somebody please advise why I'm getting too many re-renders with the below? I'm not sure why a loop is occurring when trying to setState.

```
const Product = () => {
  const router = useRouter()
  console.log({ router })

  if (typeof window === 'undefined') {
    return null;
  }

  const [item, setItem] = useState()

  const { id, matrix } = router.query
  const { data, error } = useSWR(`/api/item?itemID=${id}`)
  console.log({ data })

  if (data.Item) {
    setItem(data.Item)
  }

 return (
    <div>Test</div>
  )
}
```

**Comment by jamesmosier** [comment, marked as answer]
You are setting state if `data.Item` is present on each render. Therefore you are setting state on each render.

You could try instead to use `useEffect` to set that state if `data.Item` is present:

```js
useEffect(() => {
  if (data.Item) {
    setItem(data.Item)
  }
},[data])
```

**Reply by darrylmorley** [reply]
Your solution is better, thank you. There is no reason for a re-rendor though?

**Reply by jamesmosier** [reply]
Each time `setItem` is called the component is re-rendered.

**Reply by darrylmorley** [reply]
Thanks. 

I get that bit, I don't understand why {data} is changing and triggering setItem multiple times on page load.

**Reply by jamesmosier** [reply]
This is how the page lifecycle works:

1. Page loads, initial render happens and `useSWR` API call is executed async (the rest of the render happens while the fetch is happening).
2. `useSWR` returns `{data}`, so another render happens
3. `if (data.Item)` is now true, `setItem` is executed and another render occurs
4. `if (data.Item)` is still true so `setItem` is called and another render occurs

**Reply by darrylmorley** [reply]
Ahh, got it, being dumb! Thanks for the help.
